File: cpp/is_int.cpp
#include <iostream>
#include <functional>
#include <cmath>
#include <vector>
#include <string>
#include <limits>
#include <algorithm>
#include <unordered_map>

using namespace std::literals;
using namespace std::string_literals;

bool is_integer( double x ){
  return std::floor(x)==x;
}

bool is_integer_c( double x ){
  return std:: ceil(x)==x;
}

bool is_integer_r( double x ){
  return std::round(x)==x;
}

bool is_integer_t( double x ){
  return std::trunc(x)==x;
}

struct f_t{
  std::function<bool(double)> func;
  std::string name;
};

bool isint_floor( double x ){
  bool r = is_integer(x);
  if (
    r!=is_integer_c(x) || 
    r!=is_integer_r(x) || 
    r!=is_integer_t(x)
  ){
    throw "unepxected";
  }
  return r;
}

std::vector<f_t> FUNCS{
  f_t{ isint_floor, "floor"s },
};

struct kv_t{
  std::string name;
  double value;
};

std::vector<kv_t> KV{
  kv_t{ "0.0", 0.0 },
  kv_t{ "1.0", 1.0 },
  kv_t{ "1.5", 1.5 },
  kv_t{ "2e19", 2e19 },
  kv_t{ "9e19/7", 9e19/7 },
  kv_t{ "NAN", std::numeric_limits<double>::quiet_NaN() },
  kv_t{ "INF", std::numeric_limits<double>::infinity() },
  kv_t{ "-0.0", -0.0 },
};

int main(){
  for( auto const & f : FUNCS ){
    std::unordered_map<std::string,std::unordered_map<std::string, double>> results;
    for( auto const & kv : KV ){
      auto r = f.func(kv.value);
      results[f.name][kv.name] = r;
    }
    for( auto const & kv : KV ){
      std::cout << kv.name << "|";
      for( auto const & f : FUNCS ){
        std::cout << results[f.name][kv.name] << "|";
      }
      std::cout << std::endl;
    }
  }
  return 0;
}

File: node/is_int.js
"use strict";

function is_integer(x) {
  return x == Math.floor(x);
}

function is_int_bitor(x) {
  return x == (x|0);
}

function is_uint32_typedarray(x) {
  return (new Uint32Array([x]))[0]==x;
}

function is_uint32_compare(x) {
  return is_integer(x) && 0<=x && x<=4294967295;
}

const funcs = [
  { name: "floor", func: is_integer },
  { name: "bit or", func: is_int_bitor },
  { name: "Uint32Array", func: is_uint32_typedarray },
  { name: "uint32(比較)", func: is_uint32_compare },
];

const inputs = [
  { name: "0.0", value: 0.0 },
  { name: "1.0", value: 1.0 },
  { name: "1.5", value: 1.5 },
  { name: "1e18", value: 1e18 },
  { name: "2e19", value: 2e19 },
  { name: "NAN", value: NaN },
  { name: "INF", value: Infinity },
  { name: "-0.0", value: -0.0 },
  { name: "2147483648", value:2147483648 },
  { name: "2147483647", value:2147483647 },
  { name: "-2147483649", value:-2147483649 },
  { name: "-2147483648", value:-2147483648 },
];

let results = {};

funcs.forEach(func => {
  results[func.name] = {};
  inputs.forEach(input => {
    results[func.name][input.name] = func.func(input.value);
  });
});


console.log( "|入力|"+funcs.map( e=>{
  return e.name+"|";
}).join("") );

console.log( "|:--|"+funcs.map( ()=>{
  return ":--:|";
}).join("") );

inputs.forEach( i=>{
  console.log( "|"+i.name+"|"+funcs.map(f=>{
    return results[f.name][i.name] ? "true|" : "false|";
   }).join("") );
});